nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,83 @@
+
+This article presents basic guidelines for choosing a contract evolution strategy, avoiding common mistakes, and ensuring that contracts will be easy to evolve over time.
+
+
+TODO: link to new document
+Note: Ensure that message contracts follow the general [messages design guidelines](/nservicebus/messaging/messages-events-commands.md#designing-messages).
+
+## Adding new message contracts
+
+When adding new message contracts to a contracts assembly, update the endpoints receiving or subscribing to the new contract type first. If the sender/publisher endpoints are updated first, receivers and subscribers will not process messages for the new contract type.
+
+## Adding data to existing contracts
+
+Adding additional data to existing contracts is the most common change of contracts. There are different approaches available.
+
+### Create a new contract type
+
+1. Similar to adding a completely new contract using the original contracts name + some version postfix (e.g. `CreateOrderV2`), create a new message contract by copying the existing contract and adding the additional data to the copy.
+1. Add message handlers for the new contract type to the receiving endpoints. Keep the existing mesage handlers for the "old" contract.
+1. Update the sender/publisher to send messages using the new contract type.
+1. Once all endpoints have been updated, and no more messages using the old contract type is left in the queue, the old type and associated handlers can be safely removed.
+
+This approaches requires all receivers to be updated first, as they will be aware of the new contract type before upgrading which results in message processing failures.
+
+TODO: what about version namespaces?
+
+### Add new properties to existing contract types
+
+1. Add the new properties to the existing type.
+1. Update senders to the new contract version, setting the additional data on the message
+1. Update receivers to handle the new contract version.
+
+This approach requires less modifications to the contracts and even allows to update senders and receivers in arbitrary order as long as the receivers can handle the absence of a value for the new property.
+TODO: can serialzers handle the absence?  
+
+It is recommended to use nullable types for the new properties to allow receivers to identify whether it is dealing with an old version of the contract.
+
+### Use inheritance to create a new sub-type
+
+1. Create a new contract type, inheriting from the contract type that should be extended
+1. Add the new properties to the new sub-type
+1. Update the senders to publish/send the new sub-type
+1. Update the receivers to the new sub-type
+
+This approach requires all senders/publishers to be updated first. Due to message inheritance, receivers continue to process the original message version.
+
+## Removing contracts or properties

```

Comment:
maybe document multi-inheritance as a solution as well?
allows to have contracts with "contradicting" properties but the publisher only has to do a single publish